src/Data/Monoid/Partial.hs

```diff
@@ -0,0 +1,292 @@
+
+{- |
+A /partial monoid/ is a partial semigroup with a unit. 'pappend' with the
+unit is always defined and the identity.
+
+__Laws__
+
+@
+x `'pappend'` 'pempty'  = 'Just' x = 'pappend' 'pempty' x
+@
+
+__Homomorphisms__
+
+A /partial-monoid homomorphism/ is a partial-semigroup homomorphism which
+preserves the unit.
+-}
+class PartialSemigroup a => PartialMonoid a where
+  pempty :: a
+
+  default pempty :: Monoid a => a
+  pempty = mempty
+
+deriving newtype instance PartialSemigroup a => PartialSemigroup (Identity a)
+deriving newtype instance PartialMonoid a => PartialMonoid (Identity a)
+
+instance PartialSemigroup a => PartialSemigroup (Maybe a) where
+  x `pappend` Nothing = Just x
+  Nothing `pappend` y = Just y
+  Just x `pappend` Just y = Just <$> x `pappend` y
+-- | /Free partial monoid over a partial-semigroup/
+instance PartialSemigroup a => PartialMonoid (Maybe a) where
+  pempty = Nothing
+
+instance (PartialSemigroup a, PartialSemigroup b) => PartialSemigroup (a, b) where
+  (x1, x2) `pappend` (y1, y2) = do
+    a <- x1 `pappend` y1
+    b <- x2 `pappend` y2
+    Just (a, b)
+instance (PartialMonoid a, PartialMonoid b) => PartialMonoid (a, b) where
+  pempty = (pempty, pempty)
+
+instance (PartialSemigroup a, PartialSemigroup b, PartialSemigroup c) => PartialSemigroup (a, b, c) where
+  (x1, x2, x3) `pappend` (y1, y2, y3) = do
+    a <- x1 `pappend` y1
+    b <- x2 `pappend` y2
+    c <- x3 `pappend` y3
+    Just (a, b, c)
+instance (PartialMonoid a, PartialMonoid b, PartialMonoid c) => PartialMonoid (a, b, c) where
+  pempty = (pempty, pempty, pempty)
+
+instance (PartialSemigroup a, PartialSemigroup b) => PartialSemigroup (Either a b) where
+  Left x `pappend` Left y = Left <$> x `pappend` y
+  Right x `pappend` Right y = Right <$> x `pappend` y
+  _ `pappend` _ = Nothing
+
+instance PartialSemigroup a => PartialSemigroup (Dual a) where
+  Dual x `pappend` Dual y = Dual <$> y `pappend` x
+instance PartialMonoid a => PartialMonoid (Dual a) where
+  pempty = Dual pempty
+
+{- |
+The /free monoid over a partial semigroup/:
+
+For any partial-semigroup homomorphism @f :: a -> m@ to some monoid @m@ we
+get, 'foldMap' @f@ a monoid homomorphism and
+
+@
+'foldMap' f . 'singleton' = f
+@
+
+A value in @'FreeMonoid' S@ for some partial-semigroup @S@ is a possibly empty
+list @[s₁, s₂, …, sₙ]@ with no @sᵢ@, @sᵢ₊₁@ combinable. Thus if @S@ has no
+combinable elements these are just the finite lists of elements of @S@.
+-}
+newtype FreeMonoid a = FreeMonoid (Seq a)
+  deriving newtype (Show, Eq, Ord, Foldable, NFData)
+
+instance PartialSemigroup a => PartialSemigroup (FreeMonoid a)
+instance PartialSemigroup a => PartialMonoid (FreeMonoid a)
+
+instance PartialSemigroup a => Semigroup (FreeMonoid a) where
+  FreeMonoid xs <> FreeMonoid ys = case (viewr xs, viewl ys) of
+    (EmptyR, _) -> FreeMonoid ys
+    (_, EmptyL) -> FreeMonoid xs
+    (xs' :> x, y :< ys')
+      | Just xy <- (x `pappend` y) -> FreeMonoid $ xs' <> [xy] <> ys'
```

Comment:
Doesn't this assume that the result of combining `x` and `y` won't be combinable with the last element of `xs'` or the head element of `ys'`?  Does that follow from the partial semigroup laws?  I think I can imagine a scenario where there is a sequence of elements which are reduced as much as possible (i.e. no two consecutive elements are combinable), but appending one more element to the end triggers a "cascade" of combining where we end up with a single element at the end.